File: apps/backend/src/rhesis/backend/app/schemas/behavior.py
from datetime import datetime
from typing import List, Optional, Union

from pydantic import UUID4, Field

from rhesis.backend.app.schemas import Base
from rhesis.backend.app.schemas.tag import Tag
from rhesis.backend.app.schemas.user import UserReference


# Behavior schemas
class BehaviorBase(Base):
    name: str
    description: Optional[str] = None
    user_id: Optional[UUID4] = None
    organization_id: Optional[UUID4] = None


class BehaviorCreate(BehaviorBase):
    pass


class BehaviorUpdate(BehaviorBase):
    name: Optional[str] = None


class Behavior(BehaviorBase):
    tags: List[Tag] = Field(default_factory=list)
    created_at: Optional[Union[datetime, str]] = None
    user: Optional[UserReference] = None


# The detailed model with expanded relations.
class BehaviorDetail(Behavior):
    id: UUID4
    name: Optional[str] = None

File: apps/frontend/src/utils/api-client/interfaces/behavior.ts
import { UUID } from 'crypto';
import { TypeLookup } from './type-lookup';
import { Status } from './status';
import { User } from './user';
import { Tag } from './tag';
import { MetricScope } from './metric';

export interface BehaviorBase {
  name: string;
  description?: string | null;
  user_id?: UUID | null;
  organization_id?: UUID | null;
}

export type BehaviorCreate = BehaviorBase;

export type BehaviorUpdate = Partial<BehaviorBase>;

export interface Behavior extends BehaviorBase {
  id: UUID;
}

export interface BehaviorReference {
  id: UUID;
  name: string;
  description?: string | null;
}

export interface BehaviorWithMetrics extends BehaviorBase {
  id: UUID;
  name: string;
  description?: string | null;
  user_id?: UUID | null;
  organization_id: UUID;
  created_at?: string | null;
  user?: User | null;
  metrics: MetricWithRelationships[]; // Full metric objects with type relationships
  tags?: Tag[];
}

export interface MetricWithRelationships {
  id: UUID;
  name: string;
  description: string;
  score_type: string;
  metric_scope?: MetricScope[];

  // Required relationship objects (now always included from backend)
  metric_type: TypeLookup;
  backend_type: TypeLookup;

  // Optional relationship objects
  status?: Status | null;
}

export interface BehaviorsQueryParams {
  skip?: number;
  limit?: number;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  $filter?: string;
  include?: string;
}
